core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp

```diff
@@ -89,7 +89,11 @@ __inline__ __device__
 T atomic_fetch_sub( volatile T * const dest ,
   typename Kokkos::Impl::enable_if< sizeof(T) == sizeof(int) , const T >::type val )
 {
-  union { int i ; T t ; } oldval , assume , newval ;
+  union U {
+    int i ;
+    T t ;
+    KOKKOS_INLINE_FUNCTION U() {}
```

Comment:
The `inline` is unnecessary.  Would someone remind me why we cannot use `= default;` here?

Reply:
I was copying the  style from the other places. I am also happy to remove `inline` everywhere instead.

Have a look at https://godbolt.org/z/SRgKLm. As soon as the default constructor is non-trivial, we run into trouble.

Reply:
> I was copying the style from the other places. I am also happy to remove `inline` everywhere instead.

Consistency is good I suppose :)

> Have a look at https://godbolt.org/z/SRgKLm. As soon as the default constructor is non-trivial, we run into trouble.

Thanks.  So the problem comes from `T`'s having explicitly declared a default constructor :(

Reply:
Yes, `std::complex<>` is a more realistic example.

Reply:
While it wouldn't need to be inline you still need KOKKOS_FUNCTION for code markup.